src/Compilers/Core/Portable/DiagnosticAnalyzer/AsyncQueue.cs

```diff
@@ -300,11 +278,93 @@ private ValueTask<Optional<TElement>> TryDequeueCoreAsync()
                         _waiters = new Queue<TaskCompletionSource<Optional<TElement>>>();
                     }
 
-                    var waiter = new TaskCompletionSource<Optional<TElement>>();
+                    var waiter = new TaskCompletionSource<Optional<TElement>>(TaskCreationOptions.RunContinuationsAsynchronously);
+                    AttachCancellation(waiter, cancellationToken);
                     _waiters.Enqueue(waiter);
                     return new ValueTask<Optional<TElement>>(waiter.Task);
                 }
             }
         }
+
+        /// <summary>
+        /// Cancels a <see cref="TaskCompletionSource{TResult}.Task"/> if a given <see cref="CancellationToken"/> is canceled.
+        /// </summary>
+        /// <typeparam name="T">The type of value returned by a successfully completed <see cref="Task{TResult}"/>.</typeparam>
+        /// <param name="taskCompletionSource">The <see cref="TaskCompletionSource{TResult}"/> to cancel.</param>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/>.</param>
+        /// <seealso href="https://github.com/microsoft/vs-threading/blob/558f24c576cc620a00b20ed1fa90a5e2d13b0440/src/Microsoft.VisualStudio.Threading/ThreadingTools.cs#L181-L255"/>
+        private static void AttachCancellation<T>(TaskCompletionSource<T> taskCompletionSource, CancellationToken cancellationToken)
+        {
+            _ = taskCompletionSource ?? throw new ArgumentNullException(nameof(taskCompletionSource));
```

Comment:
not a fan of this pattern tbh :)

Reply:
This is a `private` method, would `Debug.Assert` not `throw` here.

Reply:
➡️ This is private nullable-enabled code, so I just removed the checks